src/main/java/org/hydev/mcpm/client/interaction/VariableWindowProgressSpeedCalculator.java

```diff
@@ -0,0 +1,77 @@
+package org.hydev.mcpm.client.interaction;
+
+import com.google.errorprone.annotations.Var;
+import org.checkerframework.checker.units.qual.A;
+import org.hydev.mcpm.utils.Pair;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.TreeMap;
+import java.util.TreeSet;
+
+/**
+ * Implementation of VariableWindowProgressRowBoundary using a prefix sum array.
+ */
+public class VariableWindowProgressSpeedCalculator implements VariableWindowProgressSpeedBoundary {
+    private final ArrayList<Long> psa = new ArrayList<>();
+    // psa of progress, where each element is total progress within the same millisecond
+    // milliseconds are not necessarily consecutive, for example, if no progress was made at t=22ms
+    // then psa will not contain 0 for this instance of time
+    // in other words we use coordinate compression
+
+    private final TreeMap<Long, Integer> map = new TreeMap<>(); // maps time in nanoseconds to psa indices
+    private long total = 0;
+    private long start = 0; // start time
+    private final long defaultWindow;
+
+    /**
+     * Constructs a VariableWindowProgressSpeedCalculator with the given default window for getSpeed in nanoseconds.
```

Comment:
Is such precision actually necessary? Else, things look good.

Reply:
I just want it to be consistent with the other one